src/core/net/dnssd_server.cpp

```diff
@@ -288,6 +333,15 @@ Header::Response Server::AddQuestions(const Header     &aRequestHeader,
                      response = Header::kResponseServerFailure);
     }
 
+    VerifyOrExit(numInternetDomains == 0 || numInternetDomains == aRequestHeader.GetQuestionCount(),
+                 response = Header::kResponseNameError);
+
+    aRecursiveRequired = (numInternetDomains != 0);
+    if (aRecursiveRequired)
+    {
+        VerifyOrExit(aRequestHeader.IsRecursionDesiredFlagSet(), response = Header::kResponseNameError);
```

Comment:
Suggest to add a comment describing why the RD flag is required. Maybe refers to RFC1034 that "the name server should never perform recursive service unless asked via RD". It may also be helpful of adding this requirement info to higher level comments like SetUpstreamQueryEnabled() or OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE.